File: ios/ExpoUnityModule.swift
import ExpoModulesCore

public class ExpoUnityModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoUnity")

    // Send a message from RN to Unity
    Function("postMessage") { (gameObject: String, methodName: String, message: String) in
      UnityBridge.shared().sendMessage(gameObject, methodName: methodName, message: message)
    }

    // Pause / resume Unity
    Function("pauseUnity") { (pause: Bool) in
      UnityBridge.shared().pause(pause)
    }

    // Unload Unity (free memory)
    Function("unloadUnity") {
      UnityBridge.shared().unload()
    }

    // Check if Unity is initialized
    Function("isInitialized") { () -> Bool in
      return UnityBridge.shared().isInitialized()
    }

    // Event sent from Unity → RN
    Events("onUnityMessage")

    // The native Unity view
    View(ExpoUnityView.self) {
      Events("onUnityMessage")

      Prop("autoUnloadOnUnmount") { (view: ExpoUnityView, value: Bool) in
        view.autoUnloadOnUnmount = value
      }
    }
  }
}

File: android/src/main/java/expo/modules/unity/ExpoUnityModule.kt
package expo.modules.unity

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoUnityModule : Module() {

    override fun definition() = ModuleDefinition {
        Name("ExpoUnity")

        Events("onUnityMessage")

        Function("postMessage") { gameObject: String, methodName: String, message: String ->
            UnityBridge.getInstance().sendMessage(gameObject, methodName, message)
        }

        Function("pauseUnity") { pause: Boolean ->
            UnityBridge.getInstance().setPaused(pause)
        }

        Function("unloadUnity") {
            UnityBridge.getInstance().unload()
        }

        Function("isInitialized") {
            UnityBridge.getInstance().isInitialized
        }

        View(ExpoUnityView::class) {
            Events("onUnityMessage")

            Prop("autoUnloadOnUnmount") { view: ExpoUnityView, value: Boolean ->
                view.autoUnloadOnUnmount = value
            }
        }

        OnActivityEntersBackground {
            val bridge = UnityBridge.getInstance()
            if (bridge.isInitialized) {
                bridge.wasRunningBeforeBackground = true
                bridge.setPaused(true)
            } else {
                bridge.wasRunningBeforeBackground = false
            }
        }

        OnActivityEntersForeground {
            val bridge = UnityBridge.getInstance()
            if (bridge.wasRunningBeforeBackground) {
                bridge.setPaused(false)
                bridge.wasRunningBeforeBackground = false
            }
        }

        OnActivityDestroys {
            val bridge = UnityBridge.getInstance()
            if (bridge.isInitialized) {
                bridge.unload()
            }
        }
    }
}
